---
title: "R Notebook"
output: html_notebook
---

```{r}
#import libries:
library(dplyr)
library(Seurat)
library(patchwork)
```

```{r}
# Write the data to a TSV file
df <- read.table("C:\\Users\\LENOVO\\Dropbox\\miniProject\\GSE225172_sc_MT_Babraham_Blood_aging\\sc_MT_Babraham_Blood_aging.txt", header = TRUE)
class(df)
df_methyl <- read.table("C:\\Users\\LENOVO\\Dropbox\\miniProject\\Tcells_methylation1.csv",sep=",",header = TRUE, row.names = "cellID")
print(ncol(df))
# filter df columns according to df-methyl row names
df <- df[, names(df) %in% row.names(df_methyl)]
print(ncol(df))
```
```{r}
# Assuming 'df' is your data.frame
library(Matrix)

# Convert the data.frame to a sparse matrix
md <- Matrix(nrow = count(rownames(df)), ncol = count(colnames(df)), data =as.matrix(df)[,], sparse = TRUE)
md <- as(md, "dgCMatrix") # by default, Matrix() returns dgCMatrix
class(md)# Assuming 'df' is your data.frame
```

```{r}
####mitchodrial gene
library("EnsDb.Mmusculus.v79")

# Convert from ensembl.gene to gene.symbol
geneIDsconverted <- ensembldb::select(EnsDb.Mmusculus.v79, keys= row.names(md), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
#חסרים עוד ~8,000 גנים ככל הנראה זה פסאודוגנים lncRNA
mt_genes <- geneIDsconverted[grep("^mt-", geneIDsconverted$SYMBOL), ]
```

```{r}
# Combine columns into a list
combined_list <- paste(mt_genes$SYMBOL, mt_genes$GENEID, sep = "-")
md_row=row.names(md)
for (i in 1:nrow(mt_genes)){
  for(j in 1:nrow(md)){
    if (mt_genes$GENEID[i] == md_row[[j]]){
        md_row[[j]]=combined_list[i]
    }
  }
}
rownames(md) <- md_row
```

```{r}
# Initialize the Seurat object with the raw (non-normalized data).
data1<- CreateSeuratObject(counts = md, project = "data1", min.cells = 3, min.features = 200)
data1
data1[["percent.mt"]] <- PercentageFeatureSet(data1, pattern = "^mt-")
# Visualize QC metrics as a violin plot
VlnPlot(data1, features = c("nFeature_RNA", "nCount_RNA","percent.mt"), ncol = 3)
```

```{r}
data1 <- subset(data1, subset = nFeature_RNA > 1000 & nFeature_RNA < 4500 &  nCount_RNA<750000 & percent.mt < 6)
```

```{r}
#Normalizing the data
data1 <-  SCTransform(data1, vst.flavor = "v2", verbose = FALSE)

```

```{r}

```

```{r}
#Perform linear dimensional reduction
data1 <- RunPCA(data1, features = VariableFeatures(object = data1))
# Examine and visualize PCA results a few different ways
print(data1[["pca"]], dims = 1:5, nfeatures = 5)

```

```{r}
#analysis on 2 first dims- למה רק את השתיים הראשונים???
VizDimLoadings(data1, dims = 1:2, reduction = "pca")
#נותנים יותר שונות בין התאים 
```

```{r}
#analysis the main feature
DimPlot(data1, reduction = "pca") + NoLegend()
```

```{r}
#heatmap for main first feature
DimHeatmap(data1, dims = 1, cells = 500, balanced = TRUE)
```

```{r}
#heatmap for main 15 feature- למה מדפיסים שוב עבור 15 אפשר לוותר על ההדפסה  הקודמנת???
DimHeatmap(data1, dims = 1:15, cells = 500, balanced = TRUE)
```

```{r}
#optimal number of clustering
ElbowPlot(data1)
```

```{r}
#Cluster the cells:
data1 <- FindNeighbors(data1, dims = 1:10)
data1 <- FindClusters(data1, resolution = 0.5)
# Look at cluster IDs of the first 5 cells
head(Idents(data1), 5)
```

```{r}
#Run non-linear dimensional reduction (UMAP/tSNE):
data1 <- RunUMAP(data1, dims = 1:10)
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(data1, reduction = "umap")
#save the data:
saveRDS(data1, file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\data1_tutorial.rds")
#האם ניתן לשפר את החלוקה לקלסטרים????
```

```{r}
DefaultAssay(data1) <- "SCT"
data1 <- PrepSCTFindMarkers(data1)
```

```{r}
# find all markers of cluster 0
#cluster0.markers <- FindMarkers(data1, ident.1 = 0)
#clus0 <- cluster0.markers[cluster0.markers$p_val_adj<0.05 & cluster0.markers$avg_log2FC >0,]
#row<-row.names(clus0)
#write.csv(row,file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\row0.csv")
# find all markers of cluster 1
#cluster1.markers <- FindMarkers(data1, ident.1 = 1)
#clus1<-cluster1.markers[cluster1.markers$p_val_adj<0.05 & cluster1.markers$avg_log2FC >0,]
#row<-row.names(clus1)
#write.csv(row,file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\row1.csv")
# find all markers of cluster 2
#cluster2.markers <- FindMarkers(data1, ident.1 = 2)
#clus2<-cluster2.markers[cluster2.markers$p_val_adj<0.05 & cluster2.markers$avg_log2FC >0,]
##row<-row.names(clus2)
#write.csv(row,file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\row2.csv")
# find all markers of cluster 3
#cluster3.markers <- FindMarkers(data1, ident.1 = 3)
#clus3<-cluster3.markers[cluster3.markers$p_val_adj<0.05 & cluster3.markers$avg_log2FC >0,]
#row<-row.names(clus3)
#write.csv(row,file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\row3.csv")
## find all markers of cluster 4
#cluster4.markers <- FindMarkers(data1, ident.1 = 4)
#clus4<-cluster4.markers[cluster4.markers$p_val_adj<0.05 & cluster4.markers$avg_log2FC >0,]
#row<-row.names(clus4)
#write.csv(row,file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\row4.csv")
# find all markers of cluster 0&1
#cluster01.markers <- FindMarkers(data1, ident.1 = c(0,1))
#clus01<-cluster01.markers[cluster01.markers$p_val_adj<0.05 & cluster01.markers$avg_log2FC >0,]
#row<-row.names(clus01)
#write.csv(row,file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\row01.csv")
```

```{r}
df<-read.csv("C:\\Users\\LENOVO\\Dropbox\\miniProject\\fix_annotations_ages.csv")
rownames(df) <- df$Sample.ID
row.names(df)
```

```{r}
# Merge the Seurat object with the data frame
#print meta data before 
print(data1@meta.data)
#try<- AddMetaData(try,df$cell.type,col.name ="cell_type")
#data1<- AddMetaData(data1,df$Cell.type,col.name = 'cell.type')
#data1<- AddMetaData(data1,df$Chronological_age,col.name = 'Chronological_age')
# Add 'Cell.type' metadata to the Seurat object
data1 <- AddMetaData(data1, metadata = df[, "Cell.type", drop = FALSE], col.name = "cell.type")

# Add 'Chronological_age' metadata to the Seurat object
data1 <- AddMetaData(data1, metadata = df[, "Chronological_age", drop = FALSE], col.name = "Chronological_age")
#after:
print(data1@meta.data)
```
```{r}
#save data1 after all
saveRDS(data1, file = "C:\\Users\\LENOVO\\Dropbox\\miniProject\\data1_after_all.rds")
```
```{r}
 table(data1$seurat_clusters, data1$cell.type)
```

```{r}
print(data1@meta.data)
data1 <- AddMetaData(data1, metadata = df_methyl[, "polycomb_islands_mean_methylation_values", drop = FALSE], col.name = "polycomb_islands_mean_methylation_values")
print(data1@meta.data)
```
```{r}
gene_expr <- GetAssayData(data1, layer = "data") # Extract gene expression matrix (SCT)
```
```{r}
gene_expr<-( as.data.frame(t(gene_expr)))
gene_expr$Chronological_age  <-as.vector(data1@meta.data$Chronological_age)
gene_expr$polycomb_islands_mean_methylation_values <- as.vector(data1@meta.data$polycomb_islands_mean_methylation_values)
```


```{r}
# Function to remove columns with more than 75% zero or NA values
remove_25 <- function(data, threshold = 0.75) {
  cols_to_keep <- sapply(data, function(col) {
    na_zeros_count <- sum(is.na(col) | col == 0)
    na_zeros_ratio <- na_zeros_count / length(col)
    na_zeros_ratio <= threshold
  })
  data[, cols_to_keep]
}
gene_expr <- remove_25(gene_expr)
```


```{r}
#the num of columns to see if the regression is passiable:
print(ncol(gene_expr))
# Create and fit the model
big_model <- lm(polycomb_islands_mean_methylation_values ~ ., data = gene_expr)

# Print the summary of the model
summary(big_model)
```


```{r}
# p_val df
p_val_df <- data.frame(
  Chronological_age = numeric(0),
  Methylation_percentage = numeric(0)
)

# linear regression :
for (col in colnames(gene_expr)[-((ncol(gene_expr)-1):ncol(gene_expr))]){
  model <- lm(gene_expr[[col]] ~ (gene_expr$Chronological_age + gene_expr$polycomb_islands_mean_methylation_values ))
  new_row <- data.frame(
  Chronological_age = summary(model)$coefficients["gene_expr$Chronological_age", "Pr(>|t|)"] ,
  Methylation_percentage = summary(model)$coefficients["gene_expr$polycomb_islands_mean_methylation_values", "Pr(>|t|)"],
  Estimate_methyl = ifelse(summary(model)$coefficients["gene_expr$polycomb_islands_mean_methylation_values", "Estimate"]>0, "+", "-"),
  Row_name = col
    )

   #insert to p_val_df:
    p_val_df <- rbind(p_val_df, new_row)
}
```
```{r}
p_val_df$adjust <- p.adjust(p_val_df$Methylation_percentage, method = "BH")
p_val_df$minus_log_methyl <- -log(p_val_df$adjust)
p_val_df<- p_val_df[p_val_df$adjust<0.05,]
symbol <- ensembldb::select(EnsDb.Mmusculus.v79, keys=p_val_df$Row_name, keytype = "GENEID", columns = c("SYMBOL"))
p_val_df <- merge(p_val_df, symbol, by.x = "Row_name", by.y = "GENEID", all.x = TRUE)
p_val_df$Estimate_methyl <- ifelse(p_val_df$Estimate_methyl == "+", 1, -1)
p_val_df$est_pi_log <- p_val_df$Estimate_methyl*p_val_df$minus_log_methyl
```
```{r}
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(Orthology.eg.db)

mapIt <- function(mouseids, horg, morg, orth){
    mouseg <- mapIds(morg, mouseids, "ENTREZID", "SYMBOL")
    mapped <- select(orth, mouseg, "Homo_sapiens","Mus_musculus")
    names(mapped) <- c("Mus_egid","Homo_egid")
    husymb <- select(horg, as.character(mapped[,2]), "SYMBOL","ENTREZID")
    return(Homo_symbol = husymb[,2])
}
human_gene <- mapIt(p_val_df$Row_name, org.Hs.eg.db, org.Mm.eg.db, Orthology.eg.db)
```

```{r}
# Separate the predictors (X) and the response variable (Y)
X <- as.matrix(gene_expr[, !names(gene_expr) %in% c('polycomb_islands_mean_methylation_values')])
Y <- gene_expr$polycomb_islands_mean_methylation_values
library(glmnet)

# Standardize the data (optional but recommended)
X <- scale(X)

# Create and fit the Lasso model
lasso_model <- glmnet(X, Y, alpha = 1)  # alpha = 1 for Lasso

# Perform cross-validation to find the best lambda
cv_model <- cv.glmnet(X, Y, alpha = 1)
best_lambda <- cv_model$lambda.min

# Fit the final model with the best lambda
final_model <- glmnet(X, Y, alpha = 1, lambda = best_lambda)

# Get the coefficients
coefficients <- as.matrix(coef(final_model))

# Filter and print only the non-zero coefficients
non_zero_coefficients <- coefficients[coefficients != 0, ]
print(non_zero_coefficients)


```

